Update: added operator-specific options to space-infix-ops rule

mvolkmann opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
**What rule do you want to change?**
space-infix-ops

**Does this change cause the rule to produce more or fewer warnings?**
fewer; With the new option, "x**2" would not generate a warning.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new option

**Please provide some example code that this change will affect:**

```js
<!-- example code here -->
x**2
```

**What does the rule currently do for this code?**
The rule currently reports that spaces are required around the ** operator.

**What will the rule do after it's changed?**
The rule will no longer complain about this if the noExpSpace option is used.
I'm definitely not tied to this option name. Feel free to suggest a better one.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
I added the noExpSpace to the space-infix-ops rule to allow opting out of adding space around the ** operator. This is desirable because most people are used to not having space around the exponentiation operator when writing mathematical expressions. This does not change the default behavior of this rule.

**Is there anything you'd like reviewers to focus on?**
No.

ljharb commented:
It would be good to include some examples of chained/three+ operands (`a ** b ** c` vs `a**b**c`), and using negation, like `(-a) ** b ** c` or `-(a ** b ** c)`, etc.

not-an-aardvark commented:
Thanks for the PR!

To clarify this proposal: With the `noExpSpace` option enabled, would the rule report an error when an exponentiation *does* have spaces? Or would it just tolerate either?

I'm wondering if it would be better to have a more general way to configure this rule. For example, it would be nice to be able to do something like this:

```yml
rules:
  space-infix-ops: [error, {exceptions: ['|', '**', '=']}]
```

...which would make all operators configurable, rather than just `|`.

ljharb commented:
Perhaps it would be better to change the schema to allow "always" or "never" (default "always"), and then extend the object schema to configure each operator as "always" or "never" individually?

mvolkmann commented:
@not-an-aardvark, my original intention was to tolerate either spaces or no spaces around ** when noExpSpace is specified. I like your suggestion to make all the operators configurable with an "exception" list. I also like the suggestion from @ljharb to allow configuring each operator with "always" or "never". I'll try to implement that and push changes for review.

mvolkmann commented:
@not-an-aardvark and @ljharb, I think my last commit achieves what you wanted.
You can now choose a setting for all operators and change the setting for any individual operator.
The default behavior is the same as before, except error messages are now more detailed.
I think I had to change the error messages because before it just said "Infix operator must be surrounded by spaces." Now it needs to identify the offending operator and whether it requires or does not allow spaces around it.

mvolkmann commented:
I need to update the fix code for this rule. I'm working on that now.

mvolkmann commented:
Okay, the fix for this rule should be working now. I think this is ready to review. Please let me know if I need to do anything to help get this merged.

ljharb commented on lib/rules/space-infix-ops.js:
```diff
@@ -8,6 +8,23 @@
 // Rule Definition
 //------------------------------------------------------------------------------
 
+const OPERATORS = [
```
is there any way this list could be imported from espree, so that as new operators are added, it doesn't go out of date?

mvolkmann replied:
Wouldn't we have to find a way to get that list from all parsers since we can't assume Espree is being used?

platinumazure replied:
My suggestion would be to see if a utility (e.g., ast-utils) could be augmented with this information, so at least multiple rules could benefit. But I have no issue with it being inside the rule for now.

ljharb commented on lib/rules/space-infix-ops.js:
```diff
@@ -21,33 +38,33 @@ module.exports = {
-            OPERATORS.push("**");
-        }
+        const option = context.options[0] || {};
+        const int32Hint = Boolean(option.int32Hint);
```
does eslint's codebase prefer `Boolean(x)`, or the more robust and concise `!!x`?

ljharb commented on lib/rules/space-infix-ops.js:
```diff
@@ -21,25 +38,33 @@ module.exports = {
-            "?", ":", ",", "**"
-        ];
+        const option = context.options[0] || {};
+        const int32Hint = Boolean(option.int32Hint);
```
does eslint's codebase prefer `Boolean(x)`, or the more robust and concise `!!x`?

mvolkmann replied:
I don't mind changing it and !! is clearly more concise. I'm interested in learning why it is more robust.

platinumazure replied:
Not 100% sure if this is what ljharb was getting at, but `Boolean(x)` is vulnerable to global reassignment (e.g., if someone did `Boolean = function someOtherBooleanConstructor() {}` elsewhere in the code). `!!x` is impervious to that sort of thing because operators can't be redefined in JavaScript.

ljharb replied:
Yes, that's exactly what I'm getting at. `!!` is syntax, and as such, can't be forged or tampered with.

platinumazure commented on lib/rules/space-infix-ops.js:
```diff
@@ -75,26 +116,29 @@ module.exports = {
                 node: mainNode,
                 loc: culpritToken.loc.start,
-                message: "Infix operators must be spaced.",
+                message: `${operator} operator ${must} be surrounded by spaces.`,
```
I know this will seem unintuitive in an ES6 codebase that uses template strings everywhere else, but could you please use a non-template string and `{{operator}}` and `{{must}}`, and then add a `data` object `data: { operator, must }`?

The reason for this is because I'm hoping we can eventually centralize all messages into rule metadata, and we would leave them as strings with placeholders. This is good for eventual internationalization, as well as using rule metadata to generate rules' possible messages for website searching. When we use a template string, the placeholders are substituted before we can use the string, so those other purposes would be prevented.

This only applies to `context.report()` violation messages-- please feel free to use template strings everywhere else.

See [here](http://eslint.org/docs/developer-guide/working-with-rules#contextreport) for more info.

platinumazure commented on tests/lib/rules/space-infix-ops.js:
```diff
@@ -32,15 +32,26 @@ ruleTester.run("space-infix-ops", rule, {
             output: "a + b",
             errors: [{
-                message: "Infix operators must be spaced.",
+                message: "+ operator must be surrounded by spaces.",
```
I think we should surround the operator in question with a delimiter of some sort. My preference would be backtick, but I wouldn't mind single or double quotes either.

I also think `"Operator \`+\` must be surrounded by spaces."` would be slightly more readable, just to avoid the message starting with an operator or quotes. But I'll defer to consensus on this point.

Obviously, any changes made would apply to all invalid tests.

mysticatea commented:
Thank you for contributing!

We need to update documentation.

mvolkmann commented:
@platinumazure, I added an invalid test for what I believe is the main use case for this change, preventing spaces around the ** operator.
@mysticatea, I'm updating docs/rules/space-infix-ops.md now.

not-an-aardvark commented on lib/rules/space-infix-ops.js:
```diff
@@ -8,6 +8,23 @@
+};
+
+OPERATORS.forEach(operator => {
+    schemaProperties[operator] = operatorEnum;
```
What happens if the user does something like this?

```json
{
  "rules": {
    "space-infix-ops": ["error", {"all": "never", "+": "always"}]
  }
}
```

I see two possibilities:

1. Treat that as an invalid schema
1. Disallow spaces for all operators except `+`, and require spaces for `+`

It looks like the implementation currently uses option 2, which seems good to me (it allows a user to easily set a "default" with exceptions). However, may you please add some tests to ensure that this behavior works correctly?

mvolkmann replied:
You are correct that my rule uses option 2.
I added a test and verified that this works.

mvolkmann commented:
@not-an-aardvark @ljharb @platinumazure @mysticatea 
This is ready for another review. I have updated the documentation for this rule and added an invalid test. Is there anything else I need to do?

Every time I push I get this message: "The commit summary needs to begin with a tag (such as Fix: or Update:)." But when I start a commit message with "Fix:" or "Update:" I still get the same message. Can someone tell me what I'm doing wrong?

not-an-aardvark commented on lib/rules/space-infix-ops.js:
```diff
@@ -75,26 +116,31 @@ module.exports = {
+                    fixString += sourceCode.getText(nextToken);
 
-                    return fixer.replaceText(culpritToken, fixString);
+                    return fixer.replaceText(mainNode, fixString);
```
I think this fixer will be incorrect if a value contains multiple tokens. `fixString` only contains the text from the previous token and next token, but you're replacing the entire node. As a result,

```js
foo.bar+baz
```

will get fixed to:

```js
bar + baz
```

I would recommend using `fixer.replaceTextRange` instead, and only replacing the spacing and the operator, rather than replacing the tokens around it.

```js
const fixString = shouldHaveSpaces ? ` ${culpritToken.value} ` : culpritToken.value;

return fixer.replaceTextRange([previousToken.range[1], nextToken.range[0]], fixString);
```

not-an-aardvark replied:
Also, we should make sure that the fixer doesn't cause syntax errors:

```js
/* eslint space-infix-ops: [error, {"all": "never"}] */

foo in bar;
foo instanceof bar;
foo + +bar;
foo - -bar;

// these should get fixed to

foo in bar;
foo instanceof bar;
foo+ +bar;
foo- -bar;
```

Removing *all* of the spaces in those examples would result in a SyntaxError.

To account for this, I think the fixer needs to have logic similar to the [no-extra-parens](https://github.com/eslint/eslint/blob/b7432bdea386041012052621ab66758163d58ef4/lib/rules/no-extra-parens.js#L278-L294) fixer to determine whether it can safely remove a space.

mvolkmann replied:
Good catch! I made the change you suggested and everything still works.
I also verified that "foo.bar+baz" is fixed to "foo.bar + baz" when spaces are required around +.

mvolkmann replied:
It may be a stretch for me to implement this, but I'll try. Can you point out what lines in no-extra-parens do that?

not-an-aardvark replied:
It looks like [line 283](https://github.com/eslint/eslint/blob/be739d06ec74bba1e37c1493b51ce934751ac78d/lib/rules/no-extra-parens.js#L283) of no-extra-parens checks to see if the first character of an identifier is a valid "identifier part" . For example, combining an identifier with `typeof` could accidentally create a new identifier like `typeoffoo` rather than an expression like `typeof foo`. Then the fixer inserts a space in these cases to prevent this from happening.

(Also, if you end up feeling like you're stuck or not able to implement this, feel free to let us know and we can implement it and push to your branch.)

mvolkmann replied:
I will give this a shot this coming weekend. However, I'm wondering if this might get stalled due to lack of a sponsor and thumbs up. Should I take the time to do this?

mvolkmann replied:
@not-an-aardvark, I pushed changes that add this check. I can't say I'm super confident I did what you had in mind, but the tests I added for the cases you described seem to be passing. Take a look at the diffs and feel free to improve it.

platinumazure commented:
Thanks @mvolkmann, I'll review this tomorrow or Friday. Regarding the eslintbot message, that's reading the first commit on the branch only. We can squash-merge your commits and tweak the final commit message if needed, so don't worry about the messages at this point. Sorry for the confusion.

nzakas commented:
ESLint team - we need a champion and three 👍s for this proposal to move forward.

kaicataldo commented:
I don't think this has to block this PR, but wanted to mention that [this discussion](https://github.com/eslint/eslint/issues/7489 ) is happening, though there hasn't been any movement on it for a while.

platinumazure commented:
@eslint/eslint-team Can we get a champion and one more :+1: on this?

kaicataldo commented:
Sorry @mvolkmann, looks like we lost track of this. I've put my thumb on it.

@eslint/eslint-team Looks like we just need a champion to move this forward.

mvolkmann commented:
That's okay @kaicataldo. I assumed it was because everyone is busy with holidays and has their own ESLint issues they are focused on. However, I really hope someone picks this up in January. I put a fair amount of work into this pull request and assumed based on feedback that there was support for it.

platinumazure commented:
@mvolkmann I've volunteered to champion the issue, so now this is accepted and we'll definitely get this in as quickly as we can. You may now regard me as the point contact on the ESLint team for this issue-- please feel free to ping me with any questions (but of course, also feel free to ping specific team members if your question is specific to something they said or did).

platinumazure commented on docs/rules/space-infix-ops.md:
```diff
@@ -22,21 +25,30 @@ This rule is aimed at ensuring there are spaces around infix operators.
 
-The following patterns are considered problems:
+Without options to override the default settings,
+the following patterns are considered problems:
```
On the ESLint website, we have color-coded backgrounds for code examples depending on if they're meant to show problems or correct code. Those backgrounds are generated based on the presence of certain wordings.

Basically, for problem sections, one of these prefixes should be used:

* `Examples of **incorrect** code`
* `Example of **incorrect** code`
* `Examples of additional **incorrect** code`
* `Example of addition **incorrect** code`

And you can put whatever you like after them. So for these lines, I would suggest something like: `Examples of **incorrect** code with the default settings:`

mvolkmann replied:
I copied the format I saw in the documentation for another rule. Hopefully I did what you expected.

platinumazure commented on lib/rules/space-infix-ops.js:
```diff
@@ -51,7 +53,7 @@ module.exports = {
         // Build maps of operators that must or must not
         // have spaces around them.
-        const spacesYes = {};
+        const spacesYes = {in: true, instanceof: true};
```
You'll need to rebase this (and add spaces inside the braces). We just enabled `object-curly-spacing` in the ESLint codebase.

mvolkmann replied:
I added space inside the braces.
I understand this needs to be rebased. Unfortunately rebasing has never gone well for me. Can you tell me the steps to successfully do that? The last few times I tried rebasing (as opposed to merging) the code got messed up really bad.

platinumazure commented on lib/rules/space-infix-ops.js:
```diff
@@ -89,7 +104,13 @@ module.exports = {
-                    const shouldHaveSpaces = spacesYes[value];
+
+                    // Spaces must surround the operator if either token
+                    // is not an identifier.
```
For multiline non-JSDoc comments, we prefer this style:

```js
/*
 * First line here
 * Second line here, etc., etc.
 */
```

platinumazure commented on tests/lib/rules/space-infix-ops.js:
```diff
@@ -358,6 +373,47 @@ ruleTester.run("space-infix-ops", rule, {
+            code: "a+b - c",
+            options: { all: "never", "+": "always" },
+            output: "a + b-c",
+            errors: [{
```
Hmm. Should this test have multiple errors (one for the plus token lacking spaces, one for the minus token having spaces)?

mvolkmann replied:
It makes sense to me why it should have two errors, but I don't believe it does. I didn't change that part of the code. I only changed the part that determines which operators should be processed by the existing code.

mvolkmann replied:
It makes sense to me why it should have two errors, but I don't believe it does. I didn't change that part of the code. I only changed the part that determines which operators should be processed by the existing code.

mvolkmann commented:
Thanks so much Kevin!

On Mon, Dec 19, 2016 at 9:06 AM, Kevin Partington <notifications@github.com>
wrote:

> @mvolkmann <https://github.com/mvolkmann> I've volunteered to champion
> the issue, so now this is accepted and we'll definitely get this in as
> quickly as we can. You may now regard me as the point contact on the ESLint
> team for this issue-- please feel free to ping me with any questions (but
> of course, also feel free to ping specific team members if your question is
> specific to something they said or did).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7591#issuecomment-267986706>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE10DhhllFjUvu8HOm1SlcxFn-4Rh0jks5rJp1_gaJpZM4Kwv2l>
> .
>



-- 
R. Mark Volkmann
Object Computing, Inc.

platinumazure commented:
@mvolkmann Very sorry for losing track of this. Do you still need help rebasing this?

mvolkmann commented:
Yes, that would be great. Maybe I could learn something from the experience.

On Wed, Jan 4, 2017 at 3:31 PM, Kevin Partington <notifications@github.com>
wrote:

> @mvolkmann <https://github.com/mvolkmann> Very sorry for losing track of
> this. Do you still need help rebasing this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7591#issuecomment-270492659>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE10Dz6BUG3Ho-rmOyR8chKqyA-Ozf4ks5rPA_IgaJpZM4Kwv2l>
> .
>



-- 
R. Mark Volkmann
Object Computing, Inc.

platinumazure commented:
@mvolkmann Unfortunately it seems you're making changes on your master branch-- that makes things more difficult. It's always better to create a topic branch when you're doing new work, and then only merge changes from eslint/eslint into master.

I think you *might* be able to rebase more easily if you first sever your changes off of your master branch:

1. From the latest commit you've done so far, run `git checkout -b some-branch-name`
1. `git checkout master` (to switch back to your local master). The next steps are destructive on `master`, but you should have the `some-branch-name` reference for getting back to the work you've already done.
1. `git reset --hard HEAD~20` (where 20 is the number of commits on the pull request).

Then you can merge ESLint upstream into your master:

1. `git pull --ff-only upstream master`

Then you should be able to rebase your branch on the newly synchronized master:

1. `git checkout some-branch-name`
1. `git rebase master`

Git will then attempt to replay your 20 commits using the newly synchronized `master` as a base. Every time it encounters a conflict, it will halt the process and you will need to resolve the conflict. Then you can run `git rebase --continue` to continue the process. If you think you've screwed up, run `git rebase --abort` to cancel the rebase process. But if you work through each set of conflicts carefully, you should be able to get where you need to go.

Once the rebase is complete, you'll need to push the new branch using `git push origin some-branch-name`. Then you'll need to either reconfigure this PR to use that new branch (if that's possible, I don't know if it is), or close this PR and open a new one.

If you need help, find me on Gitter: https://gitter.im/platinumazure.

Good luck!

mvolkmann commented:
I'm trying to work through your instructions. See my note below about where
I get stuck.

On Fri, Jan 13, 2017 at 10:49 AM, Kevin Partington <notifications@github.com
> wrote:

> @mvolkmann <https://github.com/mvolkmann> Unfortunately it seems you're
> making changes on your master branch-- that makes things more difficult.
> It's always better to create a topic branch when you're doing new work, and
> then only merge changes from eslint/eslint into master.
>
> I think you *might* be able to rebase more easily if you first sever your
> changes off of your master branch:
>
>    1. From the latest commit you've done so far, run git checkout -b
>    some-branch-name
>    2. git checkout master (to switch back to your local master). The next
>    steps are destructive on master, but you should have the
>    some-branch-name reference for getting back to the work you've already
>    done.
>    3. git reset --hard HEAD~20 (where 20 is the number of commits on the
>    pull request).
>
> Then you can merge ESLint upstream into your master:
>
>    1. git pull --ff-only upstream master
>
>
This gives me:

"fatal: 'upstream' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists."

Should I replace "upstream" with "origin"?

>
>
> Then you should be able to rebase your branch on the newly synchronized
> master:
>
>    1. git checkout some-branch-name
>    2. git rebase master
>
> Git will then attempt to replay your 20 commits using the newly
> synchronized master as a base. Every time it encounters a conflict, it
> will halt the process and you will need to resolve the conflict. Then you
> can run git rebase --continue to continue the process. If you think
> you've screwed up, run git rebase --abort to cancel the rebase process.
> But if you work through each set of conflicts carefully, you should be able
> to get where you need to go.
>
> Once the rebase is complete, you'll need to push the new branch using git
> push origin some-branch-name. Then you'll need to either reconfigure this
> PR to use that new branch (if that's possible, I don't know if it is), or
> close this PR and open a new one.
>
> If you need help, find me on Gitter: https://gitter.im/platinumazure.
>
> Good luck!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7591#issuecomment-272486021>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE10KC2dtPITTZdYwG0eOPLVC2vB3Meks5rR6sdgaJpZM4Kwv2l>
> .
>



-- 
R. Mark Volkmann
Object Computing, Inc.

kaicataldo commented:
@mvolkmann `upstream` refers to the master ESLint remote, while `origin` refers to your fork. So if you called `upstream` something else, you can replace it with whatever name you gave that remote. From @platinumazure's comment:

> Assumptions:

> Your fork of ESLint on GitHub is represented as a remote called origin
> The main ESLint repo on GitHub is represented as a remote called upstream.
> If you use different names, substitute them in place of "origin" and "upstream" as necessary.

Does that make sense? Let us know if anything doesn't, and definitely feel free to stop by the chat if you want some to discuss in real time: https://gitter.im/eslint/eslint

platinumazure commented:
@mvolkmann To add to @kaicataldo's [comment above](https://github.com/eslint/eslint/pull/7591#issuecomment-272658003), you can run `git remote` from your command line to see what remotes you have set up.

If you only have an `origin`, then you need to also set up an upstream remote: `git remote add upstream https://github.com/eslint/eslint.git`. This uses HTTPS, you can use Git+HTTPS or SSH if you prefer.

Generally, a best practice when developing on GitHub is to have 2 remotes configured on your local git repo: `origin` pointing to your fork, and `upstream` (or whatever other name) pointing at the base repository that you forked off of. GitHub's documentation on this is pretty light, unfortunately.

platinumazure commented:
@mvolkmann Friendly ping, anything we can do to help you out here?

not-an-aardvark commented:
@mvolkmann, are you planning finish this? ("No" is a perfectly acceptable answer, and we can pick up where you left off -- I just want to keep the process moving along.)

mvolkmann commented:
I would really love to see this get merged. I was just hung up on how to
get my branch into a good state. I understand now that I should not have
done my work on the master branch of my fork. I guess that made it hard to
deal with. I'm lacking confidence in how to correct the situation. I'd be
glad to do more work to fix it if someone can suggest what I need to do.

On Sat, Feb 25, 2017 at 4:10 PM, Teddy Katz <notifications@github.com>
wrote:

> @mvolkmann <https://github.com/mvolkmann>, are you planning finish this?
> ("No" is a perfectly acceptable answer, and we can pick up where you left
> off -- I just want to keep the process moving along.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7591#issuecomment-282516164>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE10G_1u7UZdg_LQCC7TINt7naAj1kTks5rgKbbgaJpZM4Kwv2l>
> .
>



-- 
R. Mark Volkmann
Object Computing, Inc.

not-an-aardvark commented:
If you join our [Gitter](https://gitter.im/eslint/eslint) channel, someone will probably be able to help you with git.

mvolkmann commented:
Okay @not-an-aardvark, trying Gitter now. I posted messages around 7:30 EST today. Probably everyone is still sleeping. ;-)

mvolkmann commented:
Status update:
I created a new branch for my changes and I created an upstream remote.
I ran "git rebase origin/master" and it says my current branch is up to date.
I doubt that is correct.
Can someone suggest some commands to run that can shed some light on the state of my branch?

ilyavolodin commented:
@mvolkmann `origin` is your own fork, `upstream` is main ESLint repository (if you set it up that way. To verify run `git remote -v`). What you should do is the following sequence of commands (assuming you are in the right directory and correct branch):
```sh
git fetch --all --tag // gets all changes including tags, but doesn't update your local branches
git pull --rebase upstream master // rebase current branch on top of master branch from upstream, which is ESLint repository
// if there are merge conflicts, resolve them using "git mergetool" and "git rebase --continue" commands
git push -f origin name_of_your_branch // pushes your local branch into your own fork
```

I also suggest running the same sequence from time to time on your local master branch and pushing up to your fork. That will make it easier to contribute in the future.
Let me know if you have any other questions.

mvolkmann commented:
@ilyavolodin I followed all the advice you gave above. I hope that means this is ready to merge now. Please let me know if there is anything else I need to do.
@platinumazure @not-an-aardvark

not-an-aardvark commented:
@mvolkmann That sounds great, but I don't see the new changes -- your branch looks the same as it did before. Did you remember to push the new changes to GitHub?

mvolkmann commented:
Yes, I pushed using this command:
`git push -f origin name_of_your_branch`
When I look at the `Files changed` tab of my pull request, I see the changes.

not-an-aardvark commented:
What is the name of your branch? I also see changes for this pull request, but they're from commit 6134700bc1d110e542257cdb5491d36c0348aec8, which was made back in December.

mvolkmann commented:
The name of my branch is `space-infix-opts` (should have been `ops` at the end).
I ran `git pull --rebase upstream master` while on that branch.
After that command it said I was on this branch: `(no branch, rebasing space-infix-opts)`
Then I ran `git push -f origin space-infix-opts`.

ilyavolodin commented:
@mvolkmann Looks like something went wrong, since your fork doesn't show any pushes at all. Since you created PR directly from your master, when you do a push, that push should look like this `git push -f origin master`. Also, verify that you have upstream setup. Running `git remote -v` should give you the following output:
```sh
origin  git@github.com:mvolkmann/eslint.git (fetch)
origin  git@github.com:mvolkmann/eslint.git (push)
upstream        git@github.com:eslint/eslint.git (fetch)
upstream        git@github.com:eslint/eslint.git (push)
```
Urls might not look exactly the same, if you cloned through HTTPS, and not SSH.
If all else fails, you can change permissions on your fork to allow ESLint owners to push changes to your fork. If you enable that, one of us can resolve merge conflicts and push the changes to your fork.

mvolkmann commented:
I should probably let one of the ESLint owners handle this if they can.
The `Allow edits from maintainers.` checkbox is checked.
Do I need to do anything else to give them access?

I feel bad about how much trouble this has been. I'm sure it appears that I know very little about git. In fact I use it every day, work on branches, create and merge pull requests all the time. Apparently the pattern for doing this on an open source project is a little different. One thing is for sure ... I will never again edit code on the master branch of a fork! ;-)

ilyavolodin commented:
I pushed the changes to your fork's master branch with conflicts resolved. It should be all good now (I think). Also don't worry at all about troubles with git. Git has so many different ways of doing almost the same thing. You should've seen my first open source commit and how long it took people to help me resolve it in a correct way!:-) Happens to everyone and it's not a problem at all.

mvolkmann commented:
Thanks so much @ilyavolodin!

ilyavolodin commented:
It looks like build is failing due to linting issues. Markdown issues are clear and easy to fix. But it looks like your change modified default behavior of the rule, now we are getting a lot of `space-infix-ops` linting errors throughout our codebase.

mvolkmann commented:
In the `create` method, it builds a map of the space requirements for each operator.
An operator must be surrounded by spaces if its option is set to "always" or the "all" option is set to "always" or there is no specific setting for this operator. So the default is for all operators to be surrounded by spaces. If an operator is supposed to be surrounded by spaces then `spaceYes[operator] will be set to true.

If the setting for an operator is "never"
or the "all" option is set to "never" and there is no specific setting for the operator
then `spacesNo[operator]` will be set to true.

The real checking happens on lines 92 to 104. That code determines whether each operator is surrounded by spaces (`hasSpaces`) or is not (`lacksSpaces`). The `getFirstViolatingToken` function returns the first operator it finds that has a space issue. See the logic on lines 101-104.

There is also a corresponding fix on lines 127-144.

Does anything in what I described seem suspect?

mvolkmann commented:
There are tests for these new options and they are passing for me.
Is the code where you are seeing the linting errors lacking spaces around operators?

not-an-aardvark commented:
Simplified failure case:

```js
!foo || (bar)
```

This should not cause an error to be reported, but the rule reports an error on the `(` token.

mvolkmann commented:
I see what you are seeing and I'm working on fixing it.

mvolkmann commented:
@ilyavolodin @not-an-aardvark I just pushed changes to my `master` branch that fix the problem. My fix basically involves not treating parens as operators in the rule. All the tests pass with my change (including the new test that @not-an-aardvark suggested), but it's not obvious to me whether that is an acceptable change. Let me know what you think about that.

ilyavolodin commented:
@mvolkmann It looks like there's another issue now. I've never seen this error before, but based on all of the builds breaking and testing on my local machine, it seems that Node process runs out of memory somehow, and GC doesn't seem to clean it up fast enough. I'm not sure what might be the issue, since I haven't had time yet to look at the code.

mvolkmann commented:
@platinumazure @not-an-aardvark @ilyavolodin 
Where do you guys feel we are with this PR now?
As far as I know, everything in it is working correctly.
Let me know if there is anything else you want me to do.

not-an-aardvark commented:
@mvolkmann Did you get a chance to look at my review at https://github.com/eslint/eslint/pull/7591#pullrequestreview-16612309? It describes some cases where the rule behaves incorrectly, and those cases still exist as far as I can tell.

mvolkmann commented on docs/rules/space-infix-ops.md:
```diff
@@ -20,23 +23,49 @@ While this is valid JavaScript syntax, it is hard to determine what the author i
+  can be used as an option with the values
+  `always` (require spaces), `never` (disallow spaces), or `ignore` (don't care).
+  These options override the `all` setting for that operator.
+  Options for `in` and `instanceof` are not supported because
```
Specifying `in : ignore` will have no effect.
This line makes is so regardless of the options specified, there must be spaces around the `in` and `instanceof` operators.
```js
        const spacesYes = { in: true, instanceof: true };
```

mvolkmann commented on lib/rules/space-infix-ops.js:
```diff
@@ -21,49 +40,96 @@ module.exports = {
+        function getFirstViolatingToken(left, right) {
             const tokens = sourceCode.getTokensBetween(left, right, 1);
 
             for (let i = 1, l = tokens.length - 1; i < l; ++i) {
```
I tried to make this change, but it caused an error in the tests. I think it's because this line assumes the first token is not the one be checked.
```js
                    const endOfPrev = tokens[i - 1].range[1];
```
This of course doesn't work when `i` is `0`.

not-an-aardvark replied:
Good point. Ok, this part is fine as-is then.

mvolkmann commented on lib/rules/space-infix-ops.js:
```diff
@@ -75,26 +141,30 @@ module.exports = {
-                    if (afterToken.range[0] - culpritToken.range[1] === 0) {
+                    if (shouldHaveSpaces) {
                         fixString += " ";
                     }
```
I made this change, but with the true and false values reversed.

ilyavolodin commented:
@mvolkmann Builds are failing due to Node running out of memory. I've tried this branch locally, and it's failing locally for me as well. I'm not sure why this is happening, though.

Aladdin-ADD commented:
@mvolkmann seems this PR is from your branch master. so you just push the new commit to space-infix-opts? 
would you please push to the right branch? 😄

kaicataldo commented:
@mvolkmann Are you planning to finish this up? We unfortunately can't merge this as is with the memory issue, but we'd still love your contribution!

not-an-aardvark commented:
I'm closing this for now because there hasn't been a response from the author in quite awhile, and now that we have support for selectors in core, I'm not sure this is the correct solution to the problem anymore (see https://github.com/eslint/eslint/issues/8415#issuecomment-291926241).

Although it didn't get merged, thanks very much for your work on this PR -- we really appreciate the contribution.
